File: package/main/src/Iterator/lazyTake.ts
/**
 * Lazily takes the first n values from an iterable
 * @param {Iterable<T>} iterable - The source iterable
 * @param {number} n - The number of values to take
 * @returns {Generator<T, void, undefined>} A generator yielding at most n values
 * @example
 * const first3 = lazyTake([1, 2, 3, 4, 5], 3);
 * [...first3]; // [1, 2, 3]
 */
export function* lazyTake<T>(
  iterable: Iterable<T>,
  n: number,
): Generator<T, void, undefined> {
  let count = 0;
  for (const value of iterable) {
    if (count >= n) {
      return;
    }
    yield value;
    count += 1;
  }
}

File: package/umt_rust/src/iterator/lazy_take.rs
/// Lazily takes the first n values from an iterable
///
/// # Arguments
///
/// * `iterable` - The source iterable
/// * `n` - The number of values to take
///
/// # Returns
///
/// An iterator yielding at most n values
///
/// # Example
///
/// ```
/// use umt_rust::iterator::lazy_take;
/// let first3: Vec<i32> = lazy_take(vec![1, 2, 3, 4, 5], 3).collect();
/// assert_eq!(first3, vec![1, 2, 3]);
/// ```
pub fn lazy_take<I>(iterable: I, n: usize) -> impl Iterator<Item = I::Item>
where
    I: IntoIterator,
{
    iterable.into_iter().take(n)
}
